### PR TITLE
First pass on IN/Not In

### PR DESCRIPTION
**Pull request by hntd187**
I still want to add another test or two for these (my current ones aren't sufficient), but just getting the conversation started here. This is also the first part of getting partition filter for eventually moving delta-rs over kernel partition filters.

### CODE REVIEW

File: kernel/src/engine/arrow_expression.rs
Comment (scovich) [inline kernel/src/engine/arrow_expression.rs:79]: I'm not 100% confident I understand what this code is doing but:

It looks like we need to replicate an "array scalar" value into an arrow array suitable for comparing against a batch of primitive values, right? So e.g. if I had `x in (1, 2, 3)` then `values` would be `[1, 2, 3]` and we need to explode that into `[[1, 2, 3], [1, 2, 3], ..., [1, 2, 3]]` so that the eventual `evaluate_expression` call can invoke e.g. [in_list](https://arrow.apache.org/rust/arrow/compute/kernels/comparison/fn.in_list.html)? The documentation for that function is sorely incomplete, but I guess it's marching element by element through two arrays, producing a true output element each row whose list array element contains the corresponding primitive array element? That would be general enough to handle a correlated subquery, but would be quite space-inefficient for the common case (literal in-list or uncorrelated subquery) where we compare against the same array in every row -- especially in case said array is large. 

Does arrow rust provide a "scalar" version of `in_list` that takes two primitive arrays instead? So that e.g. `scalar_in_list([1, 2, 3, 4, 5, 6], [1, 3, 5])` returns `[true, false, true, false, true, false]`?

File: kernel/src/engine/arrow_utils.rs
Comment (hntd187) [reply kernel/src/engine/arrow_utils.rs:42]: Yup, they are available in the crate for us to use, but not public otherwise.

File: kernel/src/engine/arrow_expression.rs
Comment (hntd187) [reply kernel/src/engine/arrow_expression.rs:79]: You are correct, go through and build the offsets. This approach was actually heavily inspired from how datafusion does this, so while I agree a very large list likely doesn't have great performance there isn't (to my current knowledge) a more idiomatic way to accomplish this. I generally dislike working with the list types in arrow, but to my knowledge arrow does not provide a scalar version of it. You would have to replicate the static value N number of times for each row. Primitive array, Generic string array on the left and list array on the right.

File: kernel/src/engine/arrow_expression.rs
Comment (hntd187) [reply kernel/src/engine/arrow_expression.rs:336]: The "array-ness" of the underlying physical data are called lists in arrow, which I'd imagine is meant to reduce confusion.

File: kernel/src/engine/arrow_expression.rs
Comment (scovich) [reply kernel/src/engine/arrow_expression.rs:79]: I mean really we should be build+probe a hash table for all but the smallest in-lists, to avoid paying `O(n**2)` work. 
If arrow doesn't give a way to do that maybe we need to do it ourselves?

File: kernel/src/expressions/scalars.rs
Comment (hntd187) [reply kernel/src/expressions/scalars.rs:19]: Hmmm, I see your point, but I don't have an immediate idea for what might be best here. Off the top, engine data across the board instead of introducing another opaque time I think would be a better option (assume it's reasonable achievable). 

This code was written very much to suit delta-rs's partition filtering logic which rarely has more than a few values, I've only seen it get into the thousands a few times in the wild for a hive style partition filter. But more to your other point about subqueries and BI tools generating large constant lists, I have sadly seen those hit the max length of a java array and I have seen more than a few that are in the hundreds of thousands or millions. So your concern is well placed, but perhaps given more conversation about the representation would it be acceptable to table that discussion for a subsequent PR?

File: kernel/src/engine/arrow_expression.rs
Comment (hntd187) [reply kernel/src/engine/arrow_expression.rs:390]: Good idea, done

File: kernel/src/engine/arrow_expression.rs
Comment (hntd187) [reply kernel/src/engine/arrow_expression.rs:344]: You are suggesting something very similar to what arrow itself does for like primitive types...

```rust
pub trait ArrowPrimitiveType: primitive::PrimitiveTypeSealed + 'static {
    /// Corresponding Rust native type for the primitive type.
    type Native: ArrowNativeTypeOp;

    /// the corresponding Arrow data type of this primitive type.
    const DATA_TYPE: DataType;

    /// Returns the byte width of this primitive type.
    #[deprecated(note = "Use ArrowNativeType::get_byte_width")]
    fn get_byte_width() -> usize {
        std::mem::size_of::<Self::Native>()
    }

    /// Returns a default value of this primitive type.
    ///
    /// This is useful for aggregate array ops like `sum()`, `mean()`.
    fn default_value() -> Self::Native {
        Default::default()
    }
}

mod primitive {
    pub trait PrimitiveTypeSealed {}
}```

Lifted directly from their code, I like the idea too, we will have to do a lot of this and it'll help for implementations beyond arrow to model their own engine types into ours with not a lot of hassle.

File: kernel/src/expressions/scalars.rs
Comment (scovich) [reply kernel/src/expressions/scalars.rs:19]: Yeah, we don't need to solve it immediately, but definitely worth a tracking issue because it will bite us eventually. The advantage of going engine data route is then the engine can decide how to layout the data, whether to build a hash table, etc -- which seems very in line with kernel philosophy. I don't know how easily we could abstract that away without going the engine data route.

File: kernel/src/engine/arrow_expression.rs
Comment (scovich) [reply kernel/src/engine/arrow_expression.rs:79]: Update: the hash table thing gets pretty clearly into engine territory, so if we can figure out how to offload it to engine [data], that's probably better.

File: kernel/src/expressions/scalars.rs
Comment (hntd187) [reply kernel/src/expressions/scalars.rs:19]: Not to solution too much here, but I think this would make sense as an `EngineDataLayout` trait of sorts, where certain layout specific constants are filled in. This would be somewhat of a combination of `OffsetSizeTrait` `ArrowNativeType` and `ArrowPrimitiveType` in which the end implementation denotes layout specific information about it's engine data structure. This also has some interesting outside applications from arrays for example duckdb has a row_idx type for their row IDs. it's a uint64_t for them, but for another engine this may not be the case. I think we should reasonably strive to speak the same structure of the engine when dealing with it's data, yes?

File: kernel/src/engine/arrow_expression.rs
Comment (nicklan) [inline kernel/src/engine/arrow_expression.rs:335]: Can't we optimize this quite easily?  

If `right` is a `Literal::(ArrayData)`, we can loop `left_arr` ourselves and probe `right`, possibly switching between array vs. hashtable based on the length of `left_arr`.

If `right` is a `Literal::(something_else)` I think that's an error and we can just return.

If `right` is an actual expression, we don't need to do the crazy exploding above because it'll just evaluate to _something_ and then we can use the `list_in` case.

Possibly I'm missing something, but that seems preferable to the "building of a repeated list column" above.

**Review by nicklan — CHANGES_REQUESTED** [review]
This is cool. Left a few comments.

File: kernel/src/engine/arrow_utils.rs
Comment (scovich) [inline kernel/src/engine/arrow_utils.rs:21]: this seems way too long... rustfmt?

Comment (hntd187) [reply kernel/src/engine/arrow_utils.rs:21]: rustfmt doesn't format macros, I can improve the formatting

Comment (scovich) [reply kernel/src/engine/arrow_utils.rs:21]: huh, didn't know that!

File: kernel/src/engine/arrow_expression.rs
Comment (hntd187) [reply kernel/src/engine/arrow_expression.rs:334]: You were thinking something similar to this?

```rust
match (left.as_ref(), right.as_ref()) {
  (_, Column(_)) => { // Use Arrow in_list },
  (Literal(lit), Literal(Scalar::Array(ad))) => { // Just do the comparison ourselves. },
  (_, _) => { // Some invalid comparison }
}
```

Comment (scovich) [reply kernel/src/engine/arrow_expression.rs:334]: My understanding is that we can encounter one of four cases:
* Literal value vs. array of literal values -- not really meaningful for data skipping since no column stats are implicated (and anyway the engine's query optimizer's constant expression eval should optimize it  away).
* Column vs. array of literal values -- the "standard" case, where I believe @nicklan was suggesting to manually search the array for each column value, rather than replicating the array N times.
* Column vs. uncorrelated subquery -- The engine would need to evaluate subqueries before triggering data skipping on the resulting array of literals. Might require some thought on how to integrate that with kernel's data skipping.
* Column vs. correlated subquery -- If the engine is unable to decorrelate the subquery into a join, it would need to evaluate subqueries and produce an arrow-style array of arrays. Probably not possible to integrate with data skipping, tho (because min and max column values are meaningless in this context).

Comment (nicklan) [reply kernel/src/engine/arrow_expression.rs:334]: > You were thinking something similar to this?

Yep!

Regarding Ryan's comment. I was indeed discussing the second bullet point case.  

For uncorrelated subqueries, I think the engine could/would execute them independently and then pass that as just an array of literal values in an `IN` predicate.

Correlated subqueries are obviously trickier, but as Ryan notes we probably can't use data skipping there anyway.

Comment (scovich) [reply kernel/src/engine/arrow_expression.rs:334]: Maybe I misunderstood something? This code:
```rust
(Literal(lit), Literal(Scalar::Array(ad))) => { // Just do the comparison ourselves. },
```
would correspond to my first bullet point case `(<literal scalar> IN (<... literal values ...>)`, which could be evaluated at query compile time (no need to read anything off disk, not even stats).

The second bullet point, which I suggested we might want to optimize, would be 
```rust
(Column(...), Literal(Scalar::Array(ad))) => { // Just do the comparison ourselves. },
```
To handle this case, we'd need to grab the min/max stats for the column, and check whether they bracket any of the values in the array of scalars that represents the in-list.

Comment (hntd187) [reply kernel/src/engine/arrow_expression.rs:247]: I think you understand this correctly, and I'm having a rough time coming up with a way to do column, literal that doesn't blow the code complexity up a massive amount. It's just that without some re-plumbing there doesn't look like a good way to feed column statistics into expression evaluation. Would it be acceptable to have an inefficient impl to start with and we can discuss a larger refactoring we'd might have to do to  support better handling of this?

Comment (hntd187) [reply kernel/src/engine/arrow_expression.rs:247]: @scovich @nicklan I created #291 to discuss this further

Comment (scovich) [reply kernel/src/engine/arrow_expression.rs:247]: A couple reactions:

First, data skipping support for in-list is definitely an extra piece of work outside the scope of this PR, so we can defer the necessary stats plumbing to an issue and future work.

Second: The sub-optimal evaluation is (a) in our default engine's expression eval and (b) apparently The Arrow Way ("why that way?" being an obvious but out of scope question). So I agree we can defer being more clever, or possibly even choose not to be clever at all as long as we meet the "don't pessimize" bar. 

The bigger priority IMO is ensuring that kernel's general handling of IN-list allows engines to do a better job if they are willing to implement the expression eval themselves. That part gets trickier because engines have a lot of flexibility in how they rewrite and later eval such expressions. For example, I would hope most engines are smart enough to optimize `x in (10, 11)` as `x == 10 OR x == 11`, and some engines (IBM DB2 and sqlite3 among them) choose to convert oversized in-lists into a temp relation with an equijoin clause. In both cases, kernel would never even see an in-list. And for engines that do handle (medium-sized) IN-list directly, I would expect hash tables to figure prominently for their efficient probing. So we'll need to do some thinking about what kinds of in-list eval we expect to see, and try not to accidentally block engines from using their preferred approach.

File: kernel/src/engine/arrow_expression.rs
Comment (scovich) [reply kernel/src/engine/arrow_expression.rs:256]: Is this something you hope to pursue in this PR? Or defer to an issue as future work?

File: kernel/src/engine/arrow_utils.rs
Comment (scovich) [inline kernel/src/engine/arrow_utils.rs:33]: just double checking -- the `i32` here is specializing a generic array index type, not the actual data type?

File: kernel/src/expressions/scalars.rs
Comment (scovich) [reply kernel/src/expressions/scalars.rs:19]: Thinking more about this situation -- we're choosing to model [NOT] IN as a binary expression, which requires us to introduce some new expression type that can represent the RHS. We could go one of two ways here:
1. Introduce a general array literal type, which is transparent to kernel and which [NOT] IN expressions happen to employ as their RHS.
2. Introduce a generic in-list type, which is largely opaque to kernel and which is always the RHS of a [NOT] IN expression. 

More and more, I wonder if 2/ is a better choice:
* Adding transparent array literals permanently and significantly widens the kernel API. Absent at least a second use case (preferably 3+ use cases), this seems like a lot of work and future liability for small gain?
* Forcing the engine to materialize the in-list as a rust array of (literal, type) pairs will blow up memory usage, and reduces the maximum in-list size an engine could otherwise handle. This is a "compile time" problem, independent of \what evaluation strategy the engine employs later.
* Using an opaque specific expression type for IN list allows the engine to do whatever it wants, without having to tell kernel _anything_. Kernel would just see a trait of some kind which it dutifully and blindly forwards to the engine's expression eval. We might even be able to generalize the "opaque engine expression" concept so that kernel can "handle" engine expression types it otherwise doesn't know about.

The downside of 2/ is it might make data skipping harder, since kernel doesn't have any idea what's "inside the box." It may be that we don't care, because long in-lists may turn out to not be good data skipping targets in the first place (low pruning power). If we did decide the pruning power is worth chasing, then we would need to expand the opaque in-list expression trait with method(s) that can perform data skipping. Probably very doable.

File: kernel/src/engine/arrow_expression.rs
Comment (scovich) [reply kernel/src/engine/arrow_expression.rs:247]: However, see my other comment below -- maybe we don't want kernel to expose an array literal type in the first place?

File: kernel/src/engine/arrow_utils.rs
Comment (hntd187) [reply kernel/src/engine/arrow_utils.rs:33]: Correct, this impl doesn't support large lists which are `i64` offset

File: kernel/src/engine/arrow_expression.rs
Comment (hntd187) [reply kernel/src/engine/arrow_expression.rs:256]: No, I'd hope to do this in a subsequent PR, I'd like to keep this PR simple (and perhaps a bit inefficient) because I'd imagine we have design discussions to have in order to better support this.

File: kernel/src/expressions/scalars.rs
Comment (hntd187) [reply kernel/src/expressions/scalars.rs:19]: I'd be okay with data skipping being a bit harder because another upside of 2 is that it allows us to generalize expressions in a way that allows the engine to decide what is correct. What I mean is currently there is no subset of expression functions that delta says an expression evaluator must have or what you can write in an expression. I've seen some very "interesting" expressions in the wild and making this opaque to the kernel save a trait to evaluate us would allow engines to define their own semantics here without worrying about whether it's a supported function or not. I know this doesn't have much to do with NOT/IN but just a point I figured was worth mentioning.

Comment (scovich) [reply kernel/src/expressions/scalars.rs:19]: Sounds good! Let's go with the opaque approach and worry about data skipping if/when needed

File: kernel/src/expressions/scalars.rs
Comment (hntd187) [reply kernel/src/expressions/scalars.rs:501]: I removed the arrow aspects of the test it was some left over testing I did.

File: kernel/src/expressions/scalars.rs
Comment (scovich) [reply kernel/src/expressions/scalars.rs:19]: If we _do_ eventually want to represent in-lists as opaque engine-defined expressions rather than array literals, is there a way to mark `ArrayData` as temporary/unstable/etc so we are able to remove it later without breaking the world? If nothing else, a doc comment explaining the situation might go a long way?

**Review by scovich — APPROVED** [review]
LGTM for now, assuming we can somehow retain the ability to (a) clean up the type matching; and (b) remove the `ArrayData` concept once improved in-lists of the future no longer require it.

File: kernel/src/engine/arrow_expression.rs
Comment (hntd187) [reply kernel/src/engine/arrow_expression.rs:261]: This should be handled above, LargeLists and ListViews are not supported right now on them though. (But arrow itself doesn't support views entirely right now)

Comment (scovich) [reply kernel/src/engine/arrow_expression.rs:261]: I meant, what if the right side isn't a list at all? The unconditional cast to list type would panic. Other parts of the code were using try-casting, and converting None into Err?

Comment (scovich) [reply kernel/src/engine/arrow_expression.rs:261]: Ah, I missed the newly added check that verifies the schema types. We're good!

File: kernel/src/engine/arrow_expression.rs
Comment (roeap) [reply kernel/src/engine/arrow_expression.rs]: I love the idea of capturing these mappings more formally - delta-rs rs is littered with mappings from and to arrow...

**Review by roeap — APPROVED** [review]
Very interesting discussions on this PR!!

For me it looks good, the only comment I would have, is the we may need to open a few more follow up issues to keep track. I think I saw only one mentioned, but feel there were a few more follow ups coming up in the discussions.

**Comment by roeap** [comment]
opened #314 to capture centralizing the type mappings, which I think we hadn't done yet.

**Pull request merged**
